doc/cookbook/expose-prometheus/ExposePrometheus.lhs

```diff
@@ -0,0 +1,233 @@
+  Counters
+    <$> Prometheus.register
+          (Prometheus.vector "who"
+            (Prometheus.counter
+              (Prometheus.Info "cookbook_hello" "breakdown of hello worlds")))
+    <*> Prometheus.register
+          (Prometheus.counter
+            (Prometheus.Info "cookbook_background" "number of background thread steps"))
+```
+
+We next implement the dummy "hello-world" API route.  We add a Servant type and
+implement a Servant Handler.  We want the API route to have a `who` query-param
+so that one can call `hello?who=Alice` and  get a greeting like "hello, Alice"
+in return. We use our first Prometheus counter to record how many times Alice,
+Bob, or anyone got a greeting.
+
+The handler will defaults to `n/a` as a magic value to represent the absence of
+`who` query-parameter.
+
+``` haskell
+type Greeting = Text
+
+newtype HelloWhom = HelloWhom { getWhom :: Text }
+  deriving FromHttpApiData
+
+type HelloAPI =
+  Summary "a dummy hello-world route"
+    :> "api"
+    :> "hello"
+    :> QueryParam "who" HelloWhom
+    :> Get '[JSON] Greeting
+
+-- | A function to turn an input object into a key that we use as breakdown for
+-- the `countHellos` counter. In a real-world setting you want to ponder
+-- security and privacy risks of recording user-controlled values as breakdown values.
+helloWhomToCounterBreakdown :: HelloWhom -> Text
+helloWhomToCounterBreakdown (HelloWhom txt) = txt
+
+handleHello :: Counters -> Maybe HelloWhom -> Handler Greeting
+handleHello counters Nothing = do
+  let breakdown = "n/a"
+  liftIO $ Prometheus.withLabel (countHellos counters) breakdown Prometheus.incCounter
+  pure "hello, world"
+handleHello counters (Just who) = do
+  let breakdown = helloWhomToCounterBreakdown who
+  liftIO $ Prometheus.withLabel (countHellos counters) breakdown Prometheus.incCounter
+  pure $ "hello, " <> getWhom who
+```
+
+The second metrics we use to instrument our program is a background thread
+incrementing a counter every second.
+
+``` haskell
+startBackgroundThread :: Counters -> IO ThreadId
+startBackgroundThread counters = forkIO $ forever go
+  where
+    go :: IO ()
+    go = do
+      Prometheus.incCounter (countBackground counters)
+      threadDelay 1000000
+```
+
+Now we need to implement the part where we expose the Prometheus metrics on the
+web API.
+Let's define an API route: it's a simple `HTTP GET` returning some _Metrics_.
+In this example we also add the CORS header we discussed in intro.
+
+``` haskell
+type ServePrometheusAPI =
+  Summary "Prometheus metrics"
+    :> "metrics"
+    :> Get '[PlainText]
+         (Headers '[Header "Access-Control-Allow-Origin" CORSAllowOrigin] Metrics)
+```
+
+With this API type, we now need to fill-in the blanks.  We define a `Metrics`
+object that serializes as the Prometheus text format.  We want to keep it
+simple and use `Prometheus.exportMetricsAsText` to collect the metrics as a
+text-formatted payload. Thus, our `Metrics` object contains the raw payload.
```

Comment:
+ say exportMetricsAsText is an IO object, which we need to evaluate before rendering